src/duplicacy_gcdstorage.go

```diff
@@ -83,18 +93,37 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
 		retry = err.Temporary()
 	}
 
-	if !retry || storage.backoffs[threadIndex] >= 256 {
-		storage.backoffs[threadIndex] = 1
+	if !retry || storage.backoffsRetries[threadIndex] >= MAX_NUMBER_OF_RETRIES {
+		LOG_INFO("GCD_RETRY", "Thread: %03d. Maximum number of retries reached. Backoff time: %.2f. Number of retries: %d", threadIndex, storage.backoffs[threadIndex], storage.backoffsRetries[threadIndex])
+		storage.backoffs[threadIndex] = computeInitialBackoff(minimumSleep, maximumSleep)
+		storage.backoffsRetries[threadIndex] = 0
 		return false, err
 	}
 
-	delay := float32(storage.backoffs[threadIndex]) * rand.Float32()
-	LOG_DEBUG("GCD_RETRY", "%s; retrying after %.2f seconds", message, delay)
-	time.Sleep(time.Duration(float32(storage.backoffs[threadIndex]) * float32(time.Second)))
-	storage.backoffs[threadIndex] *= 2
+	if storage.backoffs[threadIndex] < LIMIT_BACKOFF_TIME {
+		storage.backoffs[threadIndex] *= 2.0
+	} else {
+		storage.backoffs[threadIndex] = LIMIT_BACKOFF_TIME
+		storage.backoffsRetries[threadIndex] += 1
+	}
+	delay := storage.backoffs[threadIndex]*rand.Float64() + storage.backoffs[threadIndex]*rand.Float64()
+	LOG_DEBUG("GCD_RETRY", "Thread: %3d. Message: %s. Retrying after %6.2f seconds. Current backoff: %6.2f. Number of retries: %2d.", threadIndex, message, delay, storage.backoffs[threadIndex], storage.backoffsRetries[threadIndex])
+	time.Sleep(time.Duration(delay * float64(time.Second)))
+
 	return true, nil
 }
 
+/*
```

Comment:
Although I applaud the explanatory comments, I can't help feeling that most of this content is in the wrong place. In this method you've implemented an algorithm to choose a random number between min and max, but 1) there's nothing in this method about 0.1 and 0.2, that part of the comment belongs on the code that uses those values, 2) although the SO reference could be useful, you'd be better off just leaving it off and writing in plain English that you're picking a random number between min (inclusive) and max (inclusive), 3) it's not at all clear why Google won't be DDOS'ed because of the choice to sleep between min and max (or is this still part of the commentary on threads * 0.1 or 0.2, covered by my comment 1?).